pkg/api/persistentvolumeclaim/util.go

```diff
@@ -67,3 +68,20 @@ func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
 	}
 	return false
 }
+
+// NormalizeDataSources ensures that DataSource and DataSourceRef have the same contents
+// as long as both are not explicitly set.
+// This should be used by creates/gets of PVCs, but not updates
+func NormalizeDataSources(pvcSpec *core.PersistentVolumeClaimSpec) {
+	// Don't enable this behavior if the feature gate is not on
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AnyVolumeDataSource) {
+		return
+	}
+	if pvcSpec.DataSource != nil && pvcSpec.DataSourceRef == nil {
```

Comment:
This isn't clearly what the KEP calls for.  It may actually WORK but a reader of this will not comprehend it.

Where is the code that clears DataSource when it's not a PVC or Snapshot?  Should that be here?  Otherwise what happens if I send `{ DataSource = Invalid, DataSourceRef = nil }`?  Invalid will get copied to DSRef, then (somewhere) DS will be cleared, then DS and DSRef are not the same.  This converts case 5 (in KEP) into case 6.

Basically, the compat logic is so precise and subtle that I think you should centralize it and comment the heck out of it.  Take pity on the next poor sucker who deals with this.

Reply:
So the case you mention is covered by the fact that DropDisabledFields() is called before NormalizeDataSources() in PrepareForCreate() in stragtegy.go. I agree it's not easy to follow, but there are substantial tests already for the existing code, so there's a strong reason not to make a ton of changes to the existing code, and to make these new changes in a mostly additive way. I can try to address the problem with lots and lots of comments.

Reply:
By convention, the DropDisabledFields pattern holds feature-gated logic.  No feature-gates == no logic.  I'd STRONGLY encourage you to gather all of the "permanent" logic together.  Having to jump between functions to convince myself this works is unpleasant at best (and the tests are all split) and very brittle.  I can EASILY see logic around here getting "fixed" wrongly because it's not at all clear how it is supposed to work or why.